src/osd/PrimaryLogPG.cc

```diff
@@ -9924,6 +9924,60 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
   }
 }
 
+struct C_gather : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  PrimaryLogPG::OpContext *ctx;
+  C_gather(PrimaryLogPG *pg_, hobject_t oid_, PrimaryLogPG::OpContext *ctx_) : pg(pg_), oid(oid_), ctx(ctx_) {}
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    // TODO: check errors
+    pg->lock();
+    pg->cls_gather_ops.erase(oid);
+    pg->execute_ctx(ctx);
+    pg->unlock();
+  }
+};
+
+int PrimaryLogPG::cls_gather(OpContext *ctx, std::map<std::string, bufferlist*> *src_objs, const std::string& pool,
+			     const char *cls, const char *method, bufferlist& inbl)
+{
+  OpRequestRef op = ctx->op;
+  MOSDOp *m = static_cast<MOSDOp*>(op->get_nonconst_req());
+  object_locator_t oloc = m->get_object_locator();
+
+  ObjectState& obs = ctx->new_obs;
+  object_info_t& oi = obs.oi;
+  const hobject_t& soid = oi.soid;
+
+  ObjectContextRef obc = get_object_context(soid, false);
+  C_GatherBuilder gather(cct);
+
+  CLSGatherOpRef cgop(std::make_shared<CLSGatherOp>(ctx, obc, op));
```

Comment:
auto =

Reply:
Also, I don't entirely get why CLSGatherOpRef needs to be a shared_pointer.  I see only one reference owner -- cls_gather_ops, can cls_gather_ops directly contain CLSGatherOp by value?